### DISCUSSION TITLE
Custom exception handler + await request.json() == hanging service

### DISCUSSION THREAD

**Discussion by jjbankert** [fastapi/fastapi#9109] [Questions] [answered]
### Example
Run the code below and
- trigger buggy behavior: `curl -X POST "http://localhost:8000/" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{}"`
- trigger expected behavior: `curl -X POST "http://localhost:8000/" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"msg\":\"string\"}"`

```Python
import fastapi
import pydantic
from fastapi.exception_handlers import (
    request_validation_exception_handler,
)
from fastapi.exceptions import RequestValidationError

app = fastapi.FastAPI()


class BodyModel(pydantic.BaseModel):
    msg: str


@app.exception_handler(RequestValidationError)
async def log_request_validation_error(request, exception):
    try:
        body = await request.json()  # this will enter an infinite loop
        print(f"Invalid body: {body}")

    except Exception:  # this doesn't get triggered
        pass

    return await request_validation_exception_handler(request, exception)


@app.post('/')
async def endpoint(body: BodyModel, request: fastapi.Request):
    request_body = await request.json()  # just to show that parsing the request json is fine here
    print(f"valid body: {request_body}")

    return body


if __name__ == '__main__':
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description
Executing `await request.json()` in a method that's annotated `@app.exception_handler(MyException)` leads to the service hanging.

The goal of executing this function is to log the exact request parameters that lead to an error. So if there's a better way inside this exception handler to get the body parameters then that would be welcome.

### Environment

* OS: CentOS7
* FastAPI Version: encountered on 0.52.0 and 0.61.0
* Python version: 3.8

**Comment by SirTelemak** [comment]
As I can see, request getting consumed before entering `log_request_validation_error` and not caching (can't figure why yet). When you call `await request.json()` starlette waiting for input but input is already consumed that's why it hanging.
Can't say yet if it's problem with fastapi or starlette too

**Comment by SirTelemak** [comment]
Okay, as I see Starlette genereting new Request object on error handling that's why we can't access to cached json data but when we initially generating `RequestValidationError` we saving data in `RequestValidationError.body` (request.form() or request.json()) so you can access this data from `exception.body`. It looks like workaround but I can see some good solution for it rn

**Comment by jjbankert** [comment]
Ok, that work-around is good enough for now for the `RequestValidationError`. But I'm now also finding that centralized logging of `fastapi.HTTPException`s would require me to add the request body to that exception.

**Comment by ycd** [comment]
@jjbankert  I think your approach is good but could be better, body is stored inside exception.body

```
from fastapi import Request, status, FastAPI
from fastapi.responses import JSONResponse
from fastapi.exception_handlers import request_validation_exception_handler
from fastapi.exceptions import RequestValidationError
...

@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request: Request, exc: RequestValidationError):
    return JSONResponse({"Invalid body": {"body": exc.body}})

```
This will return exactly what you wanted

`{"Invalid body":{"body":{}}}`


Also for HTTPException i quite didn't understand your problem but you can go for the same approach


```
@app.exception_handler(HTTPException)
async def http_exception_handler(request, exc):
    body = jsonable_encoder(exc)
    return JSONResponse({"body": body})


@app.post("/")
async def endpoint(body: BodyModel, request: fastapi.Request):
    if body.msg == "fail this":
        raise HTTPException(status_code=404, detail=body)
    return body
```

This will return 
```
{
  "body": {
    "status_code": 404,
    "detail": {
      "msg": "fail this"
    },
    "headers": null
  }
}

```

**Comment by jjbankert** [comment]
Ahh, that does indeed work as work-around. But I'm pretty sure that always passing the request body object to methods that might trow an exception is not a desired long-term solution. It's not clear to me whether this is something that can be fixed by FastAPI or whether this needs to be solved in Starlette, in which case I wouldn't know what to exactly ask for.

**Comment by ycd** [comment]
Oh you want to send and specific message, you can send a JSONResponse to user, if your app raises 404

```python 
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from starlette.exceptions import HTTPException as StarletteHTTPException


@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request: Request, exc: StarletteHTTPException):
    return JSONResponse({"message":"endpoint not found" })
```
```python 
INFO:     127.0.0.1:48818 - "GET /not_defined_endpoint HTTP/1.1" 200 OK
Out:        "message":"endpoint not found"
```

Then if you want to include the details, you can find that inside `exc["detail"]` which contains the response body

**Comment by earien** [comment]
Is there any news on the resolution of this issue? I see there is an easy workaround in case of a `RequestValidationException` as the body is already passed, but it does not apply to custom exceptions. It would require to pass around the body content every time we raise a custom exception in the code.

**Comment by baudcode** [comment]
I have the same problem because I want to evaluate the JSON body and use custom exceptions at the same time.

**Comment by MullaRameez** [comment]
For now, I am using the below technique. Have to find a solution before deploying to production.

```
@app.post("/test2")
async def test2(request: Request):
    dt_json = await validate_request_data(request)
    return {"test" : dt_json}
    

async def validate_request_data(request):
    dt_json = {}
    req_body = await request.body()
    if not req_body:
        raise HTTPException(status_code=204, detail="No Content - Invalid Request body")
    else: 
        dt_json = await request.json()
        return dt_json
```

**Comment by juan-rybczinski** [comment]
Good solution for the case of `RequestValidationError` by `body` property.

But how about the case of `ValidationError` and `StarletteHttpException`(not raised by me)?
None of both has the `body` property.
How do I access to the request body in the exception handler at both cases?

**Comment by j0rd1smit** [comment]
Currently, this is still an issue. I want to log the request body that caused a specific exception in my application, and this is not possible using the default FastAPI error handling. My current solution is to create a [custom APIRoute](https://fastapi.tiangolo.com/advanced/custom-request-and-route/#accessing-the-request-body-in-an-exception-handler). This APIRoute looks something like:
```python
class CustomErrorHandlingRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            try:
                return await handler(request)
            except Exception as e:
                body = await request.json()
                logger.exception(f"Exception={e};RequestBody={body}")
                raise e

        return custom_route_handler
```

It works, but it would be better if the default error handling supports this.

**Comment by JustinGuese** [comment]
still a problem now

**Comment by fburgaud** [comment]
Still a problem now. Any thoughts from @tiangolo?
I'd prefer being able to do this through the exception handler (for any possible Exception, triggered by me or not) than through the APIRoute route.

**Comment by thesubuck** [comment]
@jjbankert, your fix is simple. Instead of doing ` body = await request.json() ` , use ` body = await exception.json() ` . It will work.

re-writing the "log_request_validation_error" method as

```
async def log_request_validation_error(request, exception):
    try:
        body = await exception.json()  # this will enter an infinite loop
        print(f"Invalid body: {body}")

    except Exception:  # this doesn't get triggered
        pass

    return await request_validation_exception_handler(request, exception)
```
I tested this code by not sending "msg" key in the request and I got the following error output which you intended :
```
{
	"detail": [
		{
			"loc": [
				"body",
				"msg"
			],
			"msg": "field required",
			"type": "value_error.missing"
		}
	]
}
```
@tiangolo  already documented this [in this page](https://fastapi.tiangolo.com/tutorial/handling-errors/?h=requestvalidationerror#use-the-requestvalidationerror-body) 
The reason why you could not access the "request.json()" is because the request loses its scope when exception occurs.

**Comment by caner-cetin** [comment]
Still a problem now.

**Comment by Kludex** [comment, marked as answer]
I believe this was "solved" in Starlette. FastAPI needs to bump it.

**Reply by caner-cetin** [reply]
Thanks!

**Comment by b33lz3bubTH** [comment]
i hope this gets fixed asap.

**Comment by farmovit** [comment]
Please fix it. Not usable in production otherwise.

**Reply by lsmith77** [reply]
This is a very unproductive comment.

#1 FastAPI is obviously production usable
#2 If you feel different about priorities, you can make an eloquent case or better yet offer code for a fix

**Comment by Kludex** [comment]
This is solved on the underneath library, as mentioned on https://github.com/tiangolo/fastapi/discussions/9109#discussioncomment-6959596.

Locking the issue to avoid confusion. When FastAPI bumps Starlette, it will be solved.
